sphinxdocs/private/proto_to_markdown.py

```diff
@@ -0,0 +1,429 @@
+                        f"(_default `{param.default_value}`_) "
+                        if param.default_value
+                        else "",
+                        param.doc_string if param.doc_string else "_undocumente_",
+                    ]
+                )
+            self._render_field_list(entries)
+
+        if getattr(func, "return").doc_string:
+            return_doc = _indent_block_text(getattr(func, "return").doc_string)
+            self._write(
+                _block_attrs(f"{func_anchor}_returns"),
+                "RETURNS",
+                _link_here_icon(func_anchor + "_returns"),
+                "\n",
+                ": ",
+                return_doc,
+                "\n",
+            )
+        if func.deprecated.doc_string:
+            self._write(
+                "\n\n**DEPRECATED**\n\n", func.deprecated.doc_string.strip(), "\n"
+            )
+
+    def _render_provider(self, provider):
+        self._write(
+            _block_attrs(".starlark-object"),
+            f"## {provider.provider_name}\n\n",
+        )
+
+        provider_anchor = _anchor_id(provider.provider_name)
+        self._render_signature(
+            provider.provider_name,
+            provider_anchor,
+            provider.field_info,
+            get_name=lambda f: f.name,
+        )
+
+        self._write(provider.doc_string.strip(), "\n\n")
+
+        if provider.field_info:
+            self._write(
+                _block_attrs(provider_anchor),
+                "**FIELDS** ",
+                _link_here_icon(provider_anchor + "_fields"),
+                "\n",
+                "\n",
+            )
+            entries = []
+            for field in provider.field_info:
+                entries.append(
+                    [
+                        f"{provider_anchor}_{field.name}",
+                        field.name,
+                        field.doc_string,
+                    ]
+                )
+            self._render_field_list(entries)
+
+    def _render_attributes(self, base_anchor, attributes):
+        self._write(
+            _block_attrs(f"{base_anchor}_attributes"),
+            "**ATTRIBUTES** ",
+            _link_here_icon(f"{base_anchor}_attributes"),
+            "\n",
+        )
+        entries = []
+        for attr in attributes:
+            anchor = f"{base_anchor}_{attr.name}"
+            required = "required" if attr.mandatory else "optional"
+            attr_type = self._rule_attr_type_string(attr)
+            default = f", default `{attr.default_value}`" if attr.default_value else ""
+            # The text has to be indented to be associated with the block correctly.
+            entries.append(
+                [
+                    anchor,
+                    attr.name,
+                    f"_({required} {attr_type}{default})_\n",
+                    attr.doc_string,
+                ]
+            )
+        self._render_field_list(entries)
+
+    def _render_signature(
```

Comment:
I would love to have at list a single test for each of the rendering functions to illustrate what we are doing here. That said, the tests may be too verbose, so I am OK with not having them. Especially that we are reading protos to generate the docs.

Reply:
I added a few tests to at least exercise a decent chunk of the code paths.

Testing the markdown output is a bit tough without being really brittle.

I don't really want to put too much into testing the output because after this, I'm going to try making a Sphinx plugin. That way, the markdown that is generated can be simpler (a more raw api definitions, ala what you'd do for defining a python API), and sphinx can take care of transforming it to the UI format (sphinx has a document model plus direct integration into its other facilities like cross refs etc).

Reply:
SGTM, thanks for this work!